pkg/kubelet/status/state/state_checkpoint_test.go

```diff
@@ -0,0 +1,177 @@
+				continue
+			}
+			tests = append(tests, struct {
+				name string
+				args args
+			}{
+				name: fmt.Sprintf("resource - %s%s", fact, suf),
+				args: args{
+					podResourceAllocation: PodResourceAllocation{
+						"pod1": {
+							"container1": {
+								Requests: v1.ResourceList{
+									v1.ResourceCPU:    resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+									v1.ResourceMemory: resource.MustParse(fmt.Sprintf("%s%s", fact, suf)),
+								},
+							},
+						},
+					},
+				},
+			})
+		}
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testDir := getTestDir(t)
+			originalSC, err := NewStateCheckpoint(testDir, testCheckpoint)
+			require.NoError(t, err)
+
+			err = originalSC.SetPodResourceAllocation(tt.args.podResourceAllocation)
+			require.NoError(t, err)
+
+			actual := originalSC.GetPodResourceAllocation()
+			verifyPodResourceAllocation(t, &tt.args.podResourceAllocation, &actual, "stored pod resource allocation is not equal to original pod resource allocation")
+
+			newSC, err := NewStateCheckpoint(testDir, testCheckpoint)
+			require.NoError(t, err)
+
+			actual = newSC.GetPodResourceAllocation()
+			verifyPodResourceAllocation(t, &tt.args.podResourceAllocation, &actual, "stored pod resource allocation is not equal to original pod resource allocation")
+		})
+	}
+}
+
+func Test_stateCheckpoint_formatUpgraded(t *testing.T) {
+	// Based on the PodResourceAllocationInfo struct, it's mostly possible that new field will be added
+	// in struct PodResourceAllocationInfo, rather than in struct PodResourceAllocationInfo.AllocationEntries.
+	// Emulate upgrade scenario by pretending that `ResizeStatusEntries` is a new field.
+	// The checkpoint content doesn't have it and that shouldn't prevent the checkpoint from being loaded.
+	sc := newTestStateCheckpoint(t)
+
+	// prepare old checkpoint, ResizeStatusEntries is unset,
+	// pretend that the old checkpoint is unaware for the field ResizeStatusEntries
+	checkpointContent := `{"data":"{\"allocationEntries\":{\"pod1\":{\"container1\":{\"requests\":{\"cpu\":\"1Ki\",\"memory\":\"1Ki\"}}}}}","checksum":1555601526}`
+	checkpoint := &Checkpoint{}
+	err := checkpoint.UnmarshalCheckpoint([]byte(checkpointContent))
+	require.NoError(t, err, "failed to unmarshal checkpoint")
+
+	err = sc.checkpointManager.CreateCheckpoint(sc.checkpointName, checkpoint)
+	require.NoError(t, err, "failed to create old checkpoint")
+
+	err = sc.restoreState()
+	require.NoError(t, err, "failed to restore state")
+
+	expectedPodResourceAllocationInfo := &PodResourceAllocationInfo{
+		AllocationEntries: map[string]map[string]v1.ResourceRequirements{
+			"pod1": {
+				"container1": {
+					Requests: v1.ResourceList{
+						v1.ResourceCPU:    resource.MustParse("1Ki"),
+						v1.ResourceMemory: resource.MustParse("1Ki"),
+					},
+				},
+			},
+		},
+		ResizeStatusEntries: map[string]v1.PodResizeStatus{},
+	}
```

Comment:
nit: move this to right after you define the marshalled checkpoint data, so it's easier to compare & clear that both need to be updated together.

Reply:
updated